### ISSUE TITLE
Add formatting option to remove space before a left curly brace

### ISSUE THREAD

**Issue by kendrahavens** [dotnet/roslyn#37044]
[Original developer community feedback](https://developercommunity.visualstudio.com/idea/625462/add-formatting-option-for-the-space-before-a-left.html?childToView=632562#comment-632562
)

Proposal: A check box in Tools Options and an Editorconfig rule for controlling the space before the opening curly brace when the opening curly brace is not on a newline.
- [ ] Insert space before opening curly brace when not on newline

![image](https://user-images.githubusercontent.com/8293542/60832273-51112d00-a170-11e9-8a3f-07c23f14a2da.png)

**Comment by CyrusNajmabadi** [comment]
This feels *very* non-C#y.  It would effectively be a style to go against what pretty much 99.99% (probably more) of the community does.

I think this should fall under: https://github.com/dotnet/roslyn/issues/31691

There would be a way for people to write their own custom formatting rules.  But this would not be supported as a full option out of the box.

**Comment by asdfchlwnsgy1236** [comment]
I do not know about it feeling "non-C#y", but there have been people requesting this feature since at least 2005.
Also, I do not think it would take too much effort since the option already exists for the syntactically similar C/C++.

If setting up those custom formatting rules is simple and takes little time, then I agree that will serve the purpose for the people that want this feature.

**Comment by CyrusNajmabadi** [comment]
> If setting up those custom formatting rules is simple and takes little time, then I agree that will serve the purpose for the people that want this feature.

The goal is not to provide options for every potential one-off scenario that someone wants.  That's both a maintenance cost and it greatly increases hte complexity of the implementation (rules have to figure out how they all interact with other rules).  A more flexible solution is to just make formatting something 3rd parties can extend.  Then roslyn isn't in the business of supporting every one-off formatting choice users want to make.

**Comment by jasonmalinowski** [comment]
The agreement in the design meeting was this isn't an item we would directly add support for, but should ensure that the formatting extensibility model supports implementing things like this. https://github.com/dotnet/roslyn/issues/31691 is tracking that extensibility API.

**Comment by avjoshi-07** [comment]
Looking for that space to be removed AFTER a curly bracket, something like {1, 2, 3, 4}.............instead of { 1, 2, 3, 4 }

Is that possible?

**Comment by navstev0** [comment]
> This feels _very_ non-C#y. It would effectively be a style to go against what pretty much 99.99% (probably more) of the community does.
> 
> I think this should fall under: #31691
> 
> There would be a way for people to write their own custom formatting rules. But this would not be supported as a full option out of the box.

The only reason 99.99% of the community does it, is because out of the box this is not an option. Creating a standard by not allowing options doesn't mean that it is what the community wants, it is what we are forced to deal with.

**Comment by sharwell** [comment]
> Looking for that space to be removed AFTER a curly bracket

This is a separate request, but it's also blocked on #8269.

**Comment by sharwell** [comment]
@navstev0 We don't judge common requests on usage as much as we judge it on number of requests we get. Of the formatting requests on the backlog, this one is one of the rarest requests. Normally, this would warrant a **wontfix** resolution, but we are hoping to instead take an approach of allowing certain extensibility so individuals and/or teams using these infrequently-requested styles are still able to have a first-class experience. Once #31691 is implemented, it will be possible to implement this style rule as an extension and use it in any project where the style is desired, where it will work the same way as any other rule implemented directly by the IDE.

**Comment by CyrusNajmabadi** [comment]
Tracking through https://github.com/dotnet/roslyn/issues/31691
